src/ComputeSharp.D2D1/Shaders/Interop/Effects/ResourceManagers/D2D1ResourceTextureManagerImpl.ID2D1ResourceTextureManagerInternal.cs

```diff
@@ -0,0 +1,202 @@
+            }
+
+            lock (@this->lockHandle.Target!)
+            {
+                // If the dimensions are available, validate them
+                if (dimensions is not null)
+                {
+                    uint expectedDimensions = *dimensions;
+
+                    if (@this->resourceTextureProperties.extents is not null &&
+                        @this->resourceTextureProperties.dimensions != expectedDimensions)
+                    {
+                        return E.E_INVALIDARG;
+                    }
+
+                    @this->expectedDimensions = expectedDimensions;
+                }
+
+                // If this is the first time the method is called, just store the context.
+                // Before doing so, get an ID2D1Multithread instance, to ensure thread safety.
+                // For additional info, see docs on ID2D1EffectFactoryExtensions.GetD2D1Multithread.
+                if (@this->d2D1EffectContext is null)
+                {
+                    using ComPtr<ID2D1Multithread> d2D1Multithread = default;
+
+                    int hresult = effectContext->GetD2D1Multithread(d2D1Multithread.GetAddressOf());
+
+                    // If an ID2D1Multithread object is available, we can safely store the context
+                    if (Windows.SUCCEEDED(hresult))
+                    {
+                        _ = effectContext->AddRef();
+
+                        @this->d2D1EffectContext = effectContext;
+                        @this->d2D1Multithread = d2D1Multithread.Detach();
+                    }
+
+                    return hresult;
+                }
+
+                // Otherwise, just do nothing and return S_FALSE. This allows an existing resource texture
+                // manager to be shared across effects. If the resource cannot be shared, it will just
+                // return an error when effect is actually set, with ID2D1DrawInfo::SetResourceTexture.
+                return S.S_FALSE;
+            }
+        }
+
+        /// <inheritdoc cref="ID2D1ResourceTextureManagerInternal.GetResourceTexture"/>
+        [UnmanagedCallersOnly]
+        public static int GetResourceTexture(D2D1ResourceTextureManagerImpl* @this, ID2D1ResourceTexture** resourceTexture)
+        {
+            @this = (D2D1ResourceTextureManagerImpl*)&((void**)@this)[-1];
+
+            if (*resourceTexture is not null)
+            {
+                return E.E_POINTER;
+            }
+
+            lock (@this->lockHandle.Target!)
+            {
+                // If the effect context is null, it means Initialize has not been called yet
+                if (@this->d2D1EffectContext is null)
+                {
+                    return E.E_NOT_VALID_STATE;
+                }
+
+                // If the texture has already been created, just return it
+                if (@this->d2D1ResourceTexture is not null)
+                {
+                    @this->d2D1ResourceTexture->AddRef();
```

Comment:
you need to enter the factory lock (ID2D1Multithread) before calling AddRef() on the resource texture

Reply:
This method is only called by the effect implementation (from `OnPrepareForRender`), so we already have the D2D lock. I'm not locking in either of these methods in the internal interface because of this. We only need the lock for the APIs in the public interface accessing these D2D objects, no? 🙂

Reply:
I would still do the lock. There's nothing that prevents another implementation from calling this method. Better safe than sorry.

Reply:
This is documented in the docs for the two interfaces, it is not safe to call either APIs from the internal interface directly or without already having a D2D lock. Entering a lock wouldn't help anyway, because `Initialize` gets the `ID2D1Multithread`, which of course also does a couple calls on those D2D objects (eg. `QueryInterface` on the `ID2D1EffectFactory`). If the method was called without a D2D lock, even just doing that would already not be thread-safe, so there's just no way we could properly guard against that in the first place (need a lock to be thread-safe, but getting the lock is not thread-safe).

I'll just leave this as is, because of this, the fact it's documented, and that we don't really expect anyone to implement the internal interface anyway. And if they do, they need to read the docs, otherwise it's their fault 😄